Nzedb import changes

brookesy2 opened the pull request:
Wanted to chat about this first as I will need some advice on processing sizes.

Line 101 you can modify the chunk size, I am unsure if performance would be better with larger or smaller files. Chunk size of 50 000 produced a file of around 8-10mb. 1000 is ~100kb. I processed 1000 chunks of 1000 size, took a fair whack of time, slowed down some, but not a crazy amount.

The other thing I couldn't quite get to work was processing entirely in memory, so this is probably slower than it needs to be as it writes to disk.

If I defined data = io.StringIO() then try to write out to it via pandas it would never read into COPY :( The pandas documentation says you can write to buffer http://pandas.pydata.org/pandas-docs/stable/io.html#io-store-in-csv but whenever I did the file would show up as blank to COPY. 

Also maybe sanity check my nuke values:

https://github.com/nZEDb/nZEDb/blob/ffccda0df86ce56e756a8446e3e998405ead8e7a/nzedb/controllers/PreDb.php

Oddly in the large file this magical nuke value of 69 showed up... it looks like a non nuke value though.

Error inserting into database - invalid input syntax for type boolean: "69"

Murodese commented:
Batch processing sizes really just depend on the balance of memory usage / speed. If what you're doing is pretty slow anyway, it's just balancing memory usage. The idea is that you cut down on memory usage by only loading x thousand, but not so small that you have to query the db with every single releases. 1k-10k should be fine.

Yeah, avoid disk IO at all costs. It's incredibly slow. If it's coming up blank, it's probably that you didn't .seek(0) before using the StringIO - they work identically to files, so after you've finished writing to them you need to seek back to the start so that any future read()s will actually return data.

Nuke values like that are fine, but make them an sqlalchemy enum for simplicity's sake. There should be an example in there somewhere (there is, `passworded`).

brookesy2 commented:
Thanks for the speedy response :)

If I do:

```python
		data = io.StringIO()
                data.seek(0)
		newdata.to_csv(data, index=False, header=False)
		print(newdata[:5])
		print(data.getvalue())
```

Both the print statements output values. But copy_file(engine, data, ordering, Pre) inserts 0 records.

brookesy2 commented:
@Murodese  Figured it out, it was as you said. I forgot to seek(0) before passing the value :)

Murodese commented:
Yeah, the seek(0) moves the file pointer back to the start so reads actually get data :)

brookesy2 commented:
Took me forever to actually work that out :p

brookesy2 commented:
@Murodese So its working, do you want to give it a whirl before I make a real PR? At the moment I just did a group of 100.

The only read/write from disk is now the initial gz file for the smaller imports.

Murodese commented:
Yeah, sure, I have a bunch of REQs awaiting pres. Does this PR have all the code it needs? I can just pull this into a local repo.

brookesy2 commented:
It does, but needs to be modified to default to the smaller files (defaults
to large file at the moment). This was just a PR for some questions.

Murodese commented:
Prepare a version for me to test and I'll run it against my db :)

brookesy2 commented:
Will bust it out when I get home :)
On 14 Feb 2015 15:27, "James Meneghello" <notifications@github.com> wrote:

> Prepare a version for me to test and I'll run it against my db :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Murodese/pynab/pull/160#issuecomment-74379172>.
>

brookesy2 commented:
Give that a try, I used docopt but its a bit dodgy. I still don't feel 100% with the delete/insert, but it works.

Lastly, do we want all pre's? I could easily filter away the ones that don't have requests associated.

Murodese commented:
Probably only worth getting them if they're actually useful to us, which means having reqids attached.

Murodese commented:
```
james@indexer:/opt/pynab$ python3 scripts/nzedb_pre_import.py large
Pre-Import: File predb_dump-062714.csv not found
Traceback (most recent call last):
  File "scripts/nzedb_pre_import.py", line 211, in <module>
    largeNzedbPre()
  File "scripts/nzedb_pre_import.py", line 125, in largeNzedbPre
    for chunk in dirtyChunk:
UnboundLocalError: local variable 'dirtyChunk' referenced before assignment
```

Murodese commented:
I think you forgot to actually download the file? :p

brookesy2 commented:
The large file I assume it is there.

I will fix it up today to just do requests. Will add the download too :)
On 15 Feb 2015 03:36, "James Meneghello" <notifications@github.com> wrote:

> I think you forgot to actually download the file? :p
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Murodese/pynab/pull/160#issuecomment-74402593>.
>

brookesy2 commented:
@Murodese  Give that a whirl :) I only need to update install.py to do them both now. Thing is, the big file will take forever, do you want it in install?

Murodese commented:
Yeah, just add a note that it might take a while.

Murodese commented:
Bonus points if you can put in a progress bar :3

brookesy2 commented:
I though about it but I have no idea how to estimate how many chunks per
file :) I will hit up google and see.
On 15 Feb 2015 10:44, "James Meneghello" <notifications@github.com> wrote:

> Bonus points if you can put in a progress bar :3
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Murodese/pynab/pull/160#issuecomment-74412458>.
>

Murodese commented:
Someone will have written a library to handle all that shit for you, hang on.

Murodese commented:
http://itaybb.github.io/pySmartDL/

brookesy2 commented:
Will try put it in, the DL aint too bad for me, on 100mbit lol. But I guess good for others less lucky!

I did a counter on the chunks just so people know its actually moving.

Murodese commented:
It's more annoying at 13mbps :p

I assume this worked:

```
james@indexer:/opt/pynab$ python3 scripts/nzedb_pre_import.py large
Pre-Import: File predb_dump-062714.csv not found, attempt to download - may take a while, its 300mb
Pre-Import: Extracting file
Pre-Import: Chunk import successful
```

brookesy2 commented:
Shit, I think i left a break in there while testing lol

brookesy2 commented:
Try now :)

Murodese commented:
That looks better.

Murodese commented:
```
Traceback (most recent call last):
  File "scripts/nzedb_pre_import.py", line 230, in <module>
    largeNzedbPre()
  File "scripts/nzedb_pre_import.py", line 141, in largeNzedbPre
    for chunk in dirtyChunk:
  File "/usr/local/lib/python3.4/dist-packages/pandas/io/parsers.py", line 686, in __iter__
    yield self.read(self.chunksize)
  File "/usr/local/lib/python3.4/dist-packages/pandas/io/parsers.py", line 710, in read
    ret = self._engine.read(nrows)
  File "/usr/local/lib/python3.4/dist-packages/pandas/io/parsers.py", line 1558, in read
    alldata = self._rows_to_cols(content)
  File "/usr/local/lib/python3.4/dist-packages/pandas/io/parsers.py", line 1933, in _rows_to_cols
    raise ValueError(msg)
ValueError: Expected 23 fields in line 3411901, saw 25
```

brookesy2 commented:
hrmm how odd, I let it go on for nearly a million rows. Let me fire it up again and see.

Murodese commented:
Kinda weird that if you re-run it, you get additional pres added.

brookesy2 commented:
That's what I meant earlier when it operated oddly. The process is to take a chunk, check if the name exists, if it does, remove it, then do the COPY, that's why we get no errors. What I cant figure out is, how extra ones are being added..

brookesy2 commented:
Also it seems like there might be corruption in the csv, I will add a manual check for 23 fields.

brookesy2 commented:
I'm actually unsure of the best way to do this. Unsure how the for loop operates inside of a try/catch. I guess is if it catches an error the process just stops? Which is not what we want. Maybe I need a while and a for loop. I will have a think :p

brookesy2 commented:
4 Comments in a row, talking to myself, first sign of madness. 

Pandas, being the amazing library that it is, actually has a solution for this scenario, so I have updated the code, testing it now. Added bonus, it now uses the c parser, so its even faster!

Will also add in the download meter :)

Wow, it just finished while typing this up. 700 chunks results in ~821k pres. Now that we are only importing ones with request id's its more more concise.

You will see a lot of these part way through the file:

b'Skipping line 6672155: expected 23 fields, saw 27\nSkipping line 6672156: expected 23 fields, saw 27\nSkipping line 6672157: expected 23 fields, saw 27\nSkipping line 6672160: expected 23 fields, saw 27\nSkipping line 6672161: expected 23 fields, saw 27\nSkipping line 6672162: expected 23 fields, saw 27\nSkipping line 6672168: expected 23 fields, saw 27\nSkipping line 6672169: expected 23 fields, saw 27\nSkipping line 6672369: expected 23 fields, saw 27\nSkipping line 6673188: expected 23 fields, saw 27\nSkipping line 6673327: expected 23 fields, saw 27\nSkipping line 6673331: expected 23 fields, saw 27\nSkipping line 6674174: expected 23 fields, saw 24\nSkipping line 6674545: expected 23 fields, saw 27\nSkipping line 6674695: expected 23 fields, saw 24\nSkipping line 6674955: expected 23 fields, saw 27\nSkipping line 6675774: expected 23 fields, saw 24\nSkipping line 6677034: expected 23 fields, saw 24\nSkipping line 6677120: expected 23 fields, saw 27\nSkipping line 6677738: expected 23 fields, saw 27\nSkipping line 6677739: expected 23 fields, saw 27\nSkipping line 6677740: expected 23 fields, saw 27\nSkipping line 6677741: expected 23 fields, saw 27\nSkipping line 6677743: expected 23 fields, saw 27\nSkipping line 6677744: expected 23 fields, saw 27\nSkipping line 6677745: expected 23 fields, saw 27\nSkipping line 6677746: expected 23 fields, saw 27\nSkipping line 6677747: expected 23 fields, saw 27\nSkipping line 6677748: expected 23 fields, saw 27\nSkipping line 6677749: expected 23 fields, saw 27\nSkipping line 6677750: expected 23 fields, saw 27\nSkipping line 6677751: expected 23 fields, saw 27\nSkipping line 6677752: expected 23 fields, saw 27\nSkipping line 6677753: expected 23 fields, saw 27\nSkipping line 6677754: expected 23 fields, saw 27\nSkipping line 6677755: expected 23 fields, saw 27\nSkipping line 6677756: expected 23 fields, saw 27\nSkipping line 6677757: expected 23 fields, saw 27\nSkipping line 6677758: expected 23 fields, saw 27\nSkipping line 6677759: expected 23 fields, saw 27\nSkipping line 6677760: expected 23 fields, saw 27\nSkipping line 6677761: expected 23 fields, saw 27\nSkipping line 6677762: expected 23 fields, saw 27\nSkipping line 6677763: expected 23 fields, saw 27\nSkipping line 6677764: expected 23 fields, saw 27\nSkipping line 6677765: expected 23 fields, saw 27\nSkipping line 6677766: expected 23 fields, saw 27\nSkipping line 6677777: expected 23 fields, saw 27\nSkipping line 6677787: expected 23 fields, saw 27\nSkipping line 6678037: expected 23 fields, saw 24\nSkipping line 6678155: expected 23 fields, saw 24\nSkipping line 6679036: expected 23 fields, saw 27\nSkipping line 6679040: expected 23 fields, saw 27\nSkipping line 6679183: expected 23 fields, saw 24\nSkipping line 6679672: expected 23 fields, saw 27\nSkipping line 6680835: expected 23 fields, saw 27\nSkipping line 6680920: expected 23 fields, saw 27\nSkipping line 6681348: expected 23 fields, saw 27\nSkipping line 6681741: expected 23 fields, saw 24\nSkipping line 6682168: expected 23 fields, saw 27\nSkipping line 6682363: expected 23 fields, saw 27\nSkipping line 6682416: expected 23 fields, saw 27\nSkipping line 6682900: expected 23 fields, saw 27\nSkipping line 6682921: expected 23 fields, saw 27\nSkipping line 6683104: expected 23 fields, saw 27\n'

Murodese commented:
```
Pre-Import: No pres to add from this file
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_name_key"
DETAIL:  Key (name)=(From.The.Earth.To.The.Moon.E06.FRENCH.720p.HDTV.x264-BAWLS) already exists.
CONTEXT:  COPY pres, line 1: "From.The.Earth.To.The.Moon.E06.FRENCH.720p.HDTV.x264-BAWLS,,0,TV-X264,2011-04-08 10:29:43,ZoNeNET,19..."

Pre-Import: Error processing chunk
```

Looks like it's recognising there are duplicates and then trying to add them anyway.

brookesy2 commented on scripts/nzedb_pre_import.py:
```diff
@@ -78,82 +96,143 @@ def processNzedbPre():
+			db.commit()
+		else:
+			print("Pre-Import: No pres to add from this file")
+		
```
(I wrote all of this then realized I may have found the issue!)

@Murodese Can you sanity check this? As this is the area where it messes up on the delete/insert.

169: I remove duplicates from the file itself, picking up the last row. Basically means if something was nuked/unuked we get the most recent value.

175: Removes any pres without requestids

178-88: Creates the list of names we can use in the in query on 184. These are then added to a list so that on 191 we can get the inverse and delete them.

190: Create an inverse of the list so that we can delete the results. Actually, this is where the error is I think. I don't want the inverse, I just want to delete the ones where the names are in the list. Let me try that

Murodese commented:
I left the script to run while I was at work, and holy heck is that a dirty csv.

```
Pre-Import: Error inserting into database - invalid input syntax for type boolea                                                                                                                                                                                                                                             n: "36606"
CONTEXT:  COPY pres, line 2876, column nuked: "36606"

...

Pre-Import: Error inserting into database - date/time field value out of range:                                                                                                                                                                                                                                              "0000-00-00 00:00:00"
CONTEXT:  COPY pres, line 2056, column pretime: "0000-00-00 00:00:00"
```

...etc. Might need to do some additional checking and throw shit out if it's not formatted properly. It'll be slower, but also work better (since we're not throwing out a whole chunk every time a COPY fails).

brookesy2 commented:
It is very dirty, are you using the later check ins? As there was a change that only drops rows that error. Which when I did a full import, wasn't too bad.

Murodese commented:
I last pulled like 8 hours ago, so whatever that was. Lemme take a look and debug this.

brookesy2 commented:
Ok because this:

```python

dirtyChunk = pandas.read_table('predb_dump-062714.csv', sep='\t', header=None, na_values='\\N', usecols=[0,8,10,14,16,18,20,22], names=COLNAMES, chunksize=10000, engine='c', error_bad_lines=False)
```

Should just throw up errors when a row is dodgy. I'm gonna import the whole thing and then try import over the top of it and see what comes up, as I haven't had any complaints from COPY yet.

Murodese commented:
Just gonna go through it line-by-line, waiting on pysmartdl (which, incidentally, does some interesting stuff to pycharm :))

brookesy2 commented:
Thanks for all your help mate. Sorry to put you through all this! :(

Murodese commented:
Haha, no problem. I'm just glad I don't have to write everything myself anymore <]:)

Murodese commented:
Also, relies on unix OS (gunzip is not on windows, obv). Dunno how you want to solve that - a config line underneath `unrar_path`? Make people install gunzip for windows (I assume that exists)?

brookesy2 commented:
Seems it does, but not updated since 2007.... http://gnuwin32.sourceforge.net/packages/gzip.htm

Wonder if unrar.exe can handle .gz files? It looks like winrar can. What would be your preferred path? Also how do you detect if windows? Just if gunzip fails?

Murodese commented:
`os.name` gives you a basic string: https://docs.python.org/3/library/os.html

unrar.exe can only do RAR, I'm pretty sure.

I have no easy way to push back to a PR I don't think, so I'll just list things I notice here:

`precsv.replace("'", '', inplace=True, regex=True)` should replace ' with " (doublequote), otherwise we might have \t's within fields (which will break shit). This might be where some field errors are coming from.

`if len(cleanPres) is not 0:` should use `!=`. Not sure it matters in this case, but 'is' is only for matching object references, not values.

Rest looks okay.

brookesy2 commented:
I will investigate the os.name stuff, we might have to use winrar or other application.

If I have understood you correctly on the replace, I am checking in now :)

As an aside, nearly 1m pre's with requestids, so now I will try write over the whole thing and see what errors pop out.

Murodese commented:
Not quite on that fix - should be '"' (single - double - single), ie. replace singlequotes with doublequotes (since that's what postgres expects in copy).

Yeha, 977k rows. Running postproc now to rename all my releases.

Murodese commented:
Nooooo, missing a bunch:

```
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142226
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142226                                                                                      
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142226                                                                                     
2015-02-16 16:58:49 INFO requests: found pre request id 142231 (alt.binaries.moovee) for REQ: 142231                                                          │
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142231                                                                                       │
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142231                                                                                       │
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142231                                                                                       │
2015-02-16 16:58:49 INFO requests: found pre request id 142237 (alt.binaries.moovee) for REQ: 142237                                                          │
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142237                                                                                       │
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142237                                                                                       │
2015-02-16 16:58:49 INFO requests: found pre request id 142242 (alt.binaries.moovee) for REQ: 142242 
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142242                                                                                      
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142242                                                                                      
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142242                                                                                      
2015-02-16 16:58:49 INFO requests: found pre request id 142253 (alt.binaries.moovee) for REQ: 142253
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142253                                                                                      
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142253                                                                                      
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142262                                                                                      
2015-02-16 16:58:49 INFO requests: no pre request found for REQ: 142262                                                                                      
2015-02-16 16:58:50 INFO requests: no pre request found for REQ: 142262                                                                                      
2015-02-16 16:58:50 INFO requests: found pre request id 142266 (alt.binaries.moovee) for REQ: 142266                                                       
2015-02-16 16:58:50 INFO requests: no pre request found for REQ: 142266                                                                                      
2015-02-16 16:58:50 INFO requests: no pre request found for REQ: 142266                                                                                      
2015-02-16 16:58:50 INFO requests: no pre request found for REQ: 142266                                                                                      
2015-02-16 16:58:50 INFO requests: no pre request found for REQ: 142268                                                                                      
2015-02-16 16:58:50 INFO requests: no pre request found for REQ: 142268               
```

Murodese commented:
Actually wait, it's a bunch of dupes. That's not too bad.

brookesy2 commented:
Hrmm, so basically I should be converting these singles to doubles? Sorry for being thick here lol. At the moment I replace them with nothing.

```
'The.Human.Mind.Part.3.DVDRip.x264-CBFD'		\N		\N		\N		''		'0'		\N		'X264'		'2015-02-15 05:16:39'		'#pre@corrupt'		'0'		\N
```

brookesy2 commented:
Hopefully not too many slip through!

Murodese commented:
Yeah, replace the singles with doubles. They're there so that if a \t is naturally in the data, it still knows where the column edges are. I've seen \t's in release names before, so it's probably happening again here.

brookesy2 commented:
So, if i do this:

```python
precsv.replace("'", '"', inplace=True, regex=True)
```

It doesn't error, but it also doesn't insert anything.

Murodese commented:
Incidentally, optimised requests a bit in b5d4f79d2699bf0e24eef36129707a9c40d57400 but can't test it easily. Should work, but if you have test data readily available it'd be nice.

brookesy2 commented:
@Murodese Legend! Let me have a read and learn something :)

brookesy2 commented:
Due to running the bot for a while, I only have about 5 requests not found (as I don't backfill all that far)

So it wont be much of a strain!

brookesy2 commented:
I like what you added. Makes it more generic, the group matching feature is pretty sweet too!

Looks like you switched to a query in a for loop? I wasn't sure on efficiency, but if you say its ok, all good with me! Its also much easier to read. Unfortunately due to a holdover from another language I am geared to try avoid them :)

ukharley commented:
Use python's inbuilt gzip instead of gunzip. I've tested and it works. Example below:

        inF = gzip.open('predb_dump-062714.csv.gz', 'rb')
        data = inF.read()
        inF.close()
        outF = open('predb_dump-062714.csv', 'wb')
        outF.write(data)
        outF.close()

Murodese commented:
Does it work on monstrously large files, though?

Murodese commented:
@brookesy2 basically, rather than emit two queries and have to match release names in a for loop, build the requests into a dict grouped by group name and emit one query per group. Added benefit of breaking it up a bit, so we're not loading every possible request at once.

brookesy2 commented:
@Murodese Sounds sensible to me! :)

I will try python gzip in a bit, see how it performs.

brookesy2 commented:
@Murodese @ukharley Just gave it a whirl, its fairly fast. It did hit 100% cpu for a few seconds, and fairly high mem, but I guess that is to be expected.

I just checked in relying on pandas decompression. I ran it and didn't notice any discernible difference in speed. Anyone else want to give it a try?

Murodese commented:
Pandas probably just calls Python's gzip library, so I wouldn't expect much difference.

brookesy2 commented:
@Murodese Rock on then, problem solved :)

brookesy2 commented:
This is probably merge-able unless anyone wants to test some more.

Murodese commented:
I'll merge it now. If we run into problems, that's why it's in dev.

brookesy2 commented:
You can put a black star next to my name :p

On 19 February 2015 at 14:22, James Meneghello <notifications@github.com>
wrote:

> I'll merge it now. If we run into problems, that's why it's in dev.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Murodese/pynab/pull/160#issuecomment-75059059>.
>

Murodese commented:
It's not a regular operations thing, so it's not hugely damaging if it's broken :p The worst mess I've had to clean up was one of my own making, where NZBs were being broken upon creation for about a week. The cleanup for those are a couple of alembic migrations that find those NZBs and rewrite the headers for them :p Got lucky that I didn't need any data from the binary process to fix them.

brookesy2 commented:
That was my thinking, I can hide in the non mission critical stuff!

Sounds like you have gotten into a much worse quagmire!! If you had to go
back and get the binary data again, would not have been fun. Poor guy :)

On 19 February 2015 at 14:27, James Meneghello <notifications@github.com>
wrote:

> It's not a regular operations thing, so it's not hugely damaging if it's
> broken :p The worst mess I've had to clean up was one of my own making,
> where NZBs were being broken upon creation for about a week. The cleanup
> for those are a couple of alembic migrations that find those NZBs and
> rewrite the headers for them :p Got lucky that I didn't need any data from
> the binary process to fix them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Murodese/pynab/pull/160#issuecomment-75059954>.
>
